File: app/src/main/java/com/boardgamegeek/io/ExponentialBackOff.java
package com.boardgamegeek.io;

public class ExponentialBackOff implements BackOff {
	public static final int DEFAULT_INITIAL_INTERVAL_MILLIS = 500;
	public static final double DEFAULT_RANDOMIZATION_FACTOR = 0.5;
	public static final double DEFAULT_MULTIPLIER = 1.5;
	public static final int DEFAULT_MAX_INTERVAL_MILLIS = 60000;
	public static final int DEFAULT_MAX_ELAPSED_TIME_MILLIS = 900000;

	private final int initialIntervalMillis;
	private final double randomizationFactor;
	private final double multiplier;
	private final int maxIntervalMillis;
	private final int maxElapsedTimeMillis;
	private int currentIntervalMillis;
	private long startTimeNanos;

	public ExponentialBackOff() {
		this(new Builder());
	}

	protected ExponentialBackOff(Builder builder) {
		initialIntervalMillis = builder.initialIntervalMillis;
		randomizationFactor = builder.randomizationFactor;
		multiplier = builder.multiplier;
		maxIntervalMillis = builder.maxIntervalMillis;
		maxElapsedTimeMillis = builder.maxElapsedTimeMillis;
		checkArgument(initialIntervalMillis > 0);
		checkArgument(0 <= randomizationFactor && randomizationFactor < 1);
		checkArgument(multiplier >= 1);
		checkArgument(maxIntervalMillis >= initialIntervalMillis);
		checkArgument(maxElapsedTimeMillis > 0);
		reset();
	}

	public final void reset() {
		currentIntervalMillis = initialIntervalMillis;
		startTimeNanos = System.nanoTime();
	}

	public long nextBackOffMillis() {
		if (getElapsedTimeMillis() > maxElapsedTimeMillis) {
			return STOP;
		}
		int randomizedInterval = getRandomValueFromInterval(randomizationFactor, Math.random(), currentIntervalMillis);
		incrementCurrentInterval();
		return randomizedInterval;
	}

	static int getRandomValueFromInterval(double randomizationFactor, double random, int currentIntervalMillis) {
		double delta = randomizationFactor * currentIntervalMillis;
		double minInterval = currentIntervalMillis - delta;
		double maxInterval = currentIntervalMillis + delta;
		return (int) (minInterval + (random * (maxInterval - minInterval + 1)));
	}

	public final long getElapsedTimeMillis() {
		return (System.nanoTime() - startTimeNanos) / 1000000;
	}

	private void incrementCurrentInterval() {
		// Check for overflow, if overflow is detected set the current interval to the max interval.
		if (currentIntervalMillis >= maxIntervalMillis / multiplier) {
			currentIntervalMillis = maxIntervalMillis;
		} else {
			currentIntervalMillis *= multiplier;
		}
	}

	private static void checkArgument(boolean expression) {
		if (!expression) {
			throw new IllegalArgumentException();
		}
	}

	public static class Builder {
		int initialIntervalMillis = DEFAULT_INITIAL_INTERVAL_MILLIS;
		double randomizationFactor = DEFAULT_RANDOMIZATION_FACTOR;
		double multiplier = DEFAULT_MULTIPLIER;
		int maxIntervalMillis = DEFAULT_MAX_INTERVAL_MILLIS;
		int maxElapsedTimeMillis = DEFAULT_MAX_ELAPSED_TIME_MILLIS;

		public Builder() {
		}

		public ExponentialBackOff build() {
			return new ExponentialBackOff(this);
		}

		public Builder setInitialIntervalMillis(int initialIntervalMillis) {
			this.initialIntervalMillis = initialIntervalMillis;
			return this;
		}

		public Builder setRandomizationFactor(double randomizationFactor) {
			this.randomizationFactor = randomizationFactor;
			return this;
		}

		public Builder setMultiplier(double multiplier) {
			this.multiplier = multiplier;
			return this;
		}

		public Builder setMaxIntervalMillis(int maxIntervalMillis) {
			this.maxIntervalMillis = maxIntervalMillis;
			return this;
		}

		public Builder setMaxElapsedTimeMillis(int maxElapsedTimeMillis) {
			this.maxElapsedTimeMillis = maxElapsedTimeMillis;
			return this;
		}
	}
}

File: app/src/main/java/com/boardgamegeek/io/ExponentialBackOff.kt
package com.boardgamegeek.io

import kotlin.random.Random

class ExponentialBackOff(
    private val initialIntervalMillis: Int = 500, // the number of milliseconds between the initial request and first retry
    private val randomizationFactor: Double = 0.5, // the amount of jitter in each backoff timing. (1 = zero wait time up to double)
    private val multiplier: Double = 1.5, // amount to increase the interval on each subsequent attempt (1 is same as fixed backoff)
    private val maxIntervalMillis: Int = 60_000, // maximum number of milliseconds to wait on each retry
    private val maxElapsedTimeMillis: Int = 900_000, // total milliseconds in back-off attempt
) : BackOff {
    private var currentIntervalMillis: Int = initialIntervalMillis
    private var startTimeNanos: Long = System.nanoTime()

    override fun reset() {
        currentIntervalMillis = initialIntervalMillis
        startTimeNanos = System.nanoTime()
    }

    override fun nextBackOffMillis(): Long {
        if (calculateElapsedTimeMillis() > maxElapsedTimeMillis) return BackOff.STOP

        val delta = randomizationFactor * currentIntervalMillis
        val minInterval = currentIntervalMillis - delta
        val maxInterval = currentIntervalMillis + delta
        val randomizedInterval = (minInterval + Random.nextDouble() * (maxInterval - minInterval + 1)).toInt()

        incrementCurrentInterval()
        return randomizedInterval.toLong()
    }

    private fun calculateElapsedTimeMillis() = ((System.nanoTime() - startTimeNanos) / 1_000_000L).toInt()

    private fun incrementCurrentInterval() {
        // Check for overflow, if overflow is detected set the current interval to the max interval.
        currentIntervalMillis = if (currentIntervalMillis >= maxIntervalMillis / multiplier) {
            maxIntervalMillis
        } else {
            (currentIntervalMillis * multiplier).toInt()
        }
    }
}
